File: modules/weather.py
from datetime import datetime
import requests
import logging


class WeatherData:
    def __init__(self):
        self.values = []

    def __repr__(self):
        if not self.values:
            return "Couldn't get weather"
        weather_msg = "Weather today in Berlin Köpenick:\n"
        for v in self.values:
            weather_msg += f'{v}'
        return weather_msg


def _timestamp_to_date(timestamp):
    return datetime.fromtimestamp(timestamp).time()


def _weather(token) -> WeatherData:
    # Berlin Koepenick
    lat = 52.4514534
    lon = 13.5699097
    url = f'https://api.openweathermap.org/data/2.5/forecast?lat={lat}&lon={lon}&cnt=6&units=metric&appid={token}'

    response = requests.get(url)
    weather = response.json()

    city_name = weather["city"]["name"]
    sunset_time = _timestamp_to_date(weather["city"]["sunset"])

    temp_now = weather["list"][1]["main"]["temp"]  # +3 hours ~10:00
    temp_afternoon = weather["list"][2]["main"]["temp"]  # +6 hours ~13:00
    temp_evening = weather["list"][4]["main"]["temp"]  # +12 hours ~19:00

    temp_str = f"{temp_now} - {temp_afternoon} - {temp_evening}"
    precipitation = " - ".join(set(w["main"] for entry in weather["list"] for w in entry["weather"]))

    weather_data = WeatherData()
    weather_data.values.append(temp_str)
    weather_data.values.append(precipitation)
    weather_data.values.append(f"Sunset at {sunset_time}")
    return weather_data


def weather(token) -> WeatherData:
    try:
        return _weather(token)
    except:
        logging.exception("Couldn't get weather")
        return "Couldn't get weather"  # TODO fix

File: gorefactoring/modules/weather.go
package modules

import (
	"fmt"
	"log"
	utils "main/utils"
	"strings"
	"time"
)


func GetWeather(token string) *WeatherData {
	res, err := getWeather(token)
    if err != nil {
		log.Printf("Error in Weather module: %s", err)
	 }
	return res
}

func getWeather(token string) (*WeatherData, error) {
	// Berlin Koepenick
	lat := 52.4514534
	lon := 13.5699097
	url := fmt.Sprintf("https://api.openweathermap.org/data/2.5/forecast?lat=%f&lon=%f&cnt=6&units=metric&appid=%s", lat, lon, token)

	weatherResponse := weatherResponse{}
	err := utils.DoGet(url, &weatherResponse)
	if err != nil {
		return nil, err
	}

	return weatherResponse.GetTodayWeather(), nil
}

type weatherResponse struct {
	City struct {
		Name  string `json:"name"`
		Sunset int64  `json:"sunset"`
	} `json:"city"`
	List []struct {
		Main struct {
			Temp float64 `json:"temp"`
		} `json:"main"`
		Weather []struct {
			Main string `json:"main"`
		} `json:"weather"`
	} `json:"list"`
}

type WeatherData struct {
	Temp string
	Precipitation string
	SunsetTime string
}

func (w *WeatherData) String() string {
	if w == nil {
		return ""
	}
	weatherMsg := "Weather today in Berlin Köpenick:\n"
	weatherMsg += fmt.Sprintf("%s\n%s\nSunset at %s", w.Temp, w.Precipitation, w.SunsetTime)
	return weatherMsg
}

func (w weatherResponse) GetTodayWeather() *WeatherData {
	tempNow := w.List[1].Main.Temp // +3 hours ~10:00
	tempAfternoon := w.List[2].Main.Temp // +6 hours ~13:00
	tempEvening := w.List[4].Main.Temp // +12 hours ~19:00

	tempStr := fmt.Sprintf("%.2f : %.2f : %.2f", tempNow, tempAfternoon, tempEvening)

	var precipitation []string
	for _, entry := range w.List {
		for _, weather := range entry.Weather {
			precipitation = append(precipitation, weather.Main)
		}
	}
	precipitation = utils.RemoveDuplicate(precipitation)
	precipitationStr := strings.Join(precipitation, " - ")

	sunsetTime := time.Unix(w.City.Sunset, 0).Format("15:04")

	return &WeatherData{tempStr, precipitationStr, sunsetTime}
}